File: src/main/java/no/nav/fo/veilarbregistrering/arbeidssoker/adapter/ArenaOrdsTokenProviderClient.java
package no.nav.fo.veilarbregistrering.arbeidssoker.adapter;

import com.fasterxml.jackson.annotation.JsonAlias;
import no.nav.common.rest.client.RestClient;
import no.nav.common.rest.client.RestUtils;
import okhttp3.*;

import java.io.IOException;
import java.time.LocalDateTime;

import static java.time.temporal.ChronoUnit.SECONDS;
import static no.nav.common.utils.AuthUtils.basicCredentials;
import static no.nav.common.utils.EnvironmentUtils.getRequiredProperty;
import static org.springframework.http.HttpHeaders.AUTHORIZATION;
import static org.springframework.http.HttpHeaders.CACHE_CONTROL;

public class ArenaOrdsTokenProviderClient {

    public static final String ARENA_ORDS_CLIENT_ID_PROPERTY = "ARENA_ORDS_CLIENT_ID";
    public static final String ARENA_ORDS_CLIENT_SECRET_PROPERTY = "ARENA_ORDS_CLIENT_SECRET";

    private static final int MINIMUM_TIME_TO_EXPIRE_BEFORE_REFRESH = 60;

    private final OkHttpClient client;

    private final String arenaOrdsUrl;

    public ArenaOrdsTokenProviderClient(String arenaOrdsUrl) {
        this(arenaOrdsUrl, RestClient.baseClient());
    }

    public ArenaOrdsTokenProviderClient(String arenaOrdsUrl, OkHttpClient client) {
        this.arenaOrdsUrl = arenaOrdsUrl;
        this.client = client;
    }

    private TokenCache tokenCache = null;

    public String getToken() {
        if (tokenIsSoonExpired()) {
            refreshToken();
        }
        return tokenCache.getOrdsToken().getAccessToken();
    }

    private void refreshToken() {
        String basicAuth = basicCredentials(
                getRequiredProperty(ARENA_ORDS_CLIENT_ID_PROPERTY),
                getRequiredProperty(ARENA_ORDS_CLIENT_SECRET_PROPERTY));

        Request request = new Request.Builder()
                .url(arenaOrdsUrl)
                .header(CACHE_CONTROL, "no-cache")
                .header(AUTHORIZATION, basicAuth)
                .post(RequestBody.create(MediaType.get("application/x-www-form-urlencoded"), "grant_type=client_credentials"))
                .build();

        try (Response response = client.newCall(request).execute()) {
            RestUtils.throwIfNotSuccessful(response);
            OrdsToken ordsToken = RestUtils.parseJsonResponseOrThrow(response, OrdsToken.class);
            this.tokenCache = new TokenCache(ordsToken);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    private boolean tokenIsSoonExpired() {
        return tokenCache == null || timeToRefresh().isBefore(LocalDateTime.now());
    }

    private LocalDateTime timeToRefresh() {
        return tokenCache.getTime().plus(
                tokenCache.getOrdsToken().getExpiresIn() - MINIMUM_TIME_TO_EXPIRE_BEFORE_REFRESH, SECONDS);
    }

    private static class TokenCache {
        private final OrdsToken ordsToken;
        private final LocalDateTime time;

        TokenCache(OrdsToken ordsToken) {
            this.ordsToken = ordsToken;
            this.time = LocalDateTime.now();
        }

        public OrdsToken getOrdsToken() {
            return this.ordsToken;
        }

        public LocalDateTime getTime() {
            return this.time;
        }
    }

    static class OrdsToken {

        @JsonAlias("access_token")
        String accessToken;

        @JsonAlias("token_type")
        String tokenType;

        @JsonAlias("expires_in")
        int expiresIn;

        public OrdsToken() {
            //default constructor for `com.fasterxml.jackson.databind` som benyttes for å deserialisere.
        };

        public OrdsToken(String accessToken, String tokenType, int expiresIn) {
            this.accessToken = accessToken;
            this.tokenType = tokenType;
            this.expiresIn = expiresIn;
        }

        public String getAccessToken() {
            return accessToken;
        }

        public int getExpiresIn() {
            return expiresIn;
        }
    }
}

File: src/main/kotlin/no/nav/fo/veilarbregistrering/arbeidssoker/adapter/ArenaOrdsTokenProviderClient.kt
package no.nav.fo.veilarbregistrering.arbeidssoker.adapter

import com.fasterxml.jackson.annotation.JsonAlias
import com.fasterxml.jackson.module.kotlin.readValue
import no.nav.fo.veilarbregistrering.config.isOnPrem
import no.nav.fo.veilarbregistrering.config.objectMapper
import no.nav.fo.veilarbregistrering.config.requireProperty
import no.nav.fo.veilarbregistrering.http.defaultHttpClient
import okhttp3.HttpUrl
import okhttp3.MediaType
import okhttp3.Request
import okhttp3.RequestBody
import org.springframework.http.HttpHeaders
import java.io.IOException
import java.time.LocalDateTime
import java.time.temporal.ChronoUnit
import java.util.*

class ArenaOrdsTokenProviderClient(
    private val arenaOrdsUrl: String,
    private val proxyTokenProvider: () -> String
) {

    private var tokenCache: TokenCache? = null
    val token: String
        get() {
            if (tokenIsSoonExpired()) {
                tokenCache = TokenCache(getRefreshedToken())
            }
            return tokenCache!!.ordsToken.accessToken
        }

    private fun getRefreshedToken(): OrdsToken {
        val request = if (isOnPrem()) buildRequest() else buildRequestGcp()
        return try {
            defaultHttpClient().newCall(request).execute().use { response ->
                when {
                    response.isSuccessful -> {
                        response.body()?.string()
                            ?.let { bodyString -> objectMapper.readValue<OrdsToken>(bodyString) }
                            ?: throw IOException("Token response body was null")
                    }
                    else -> throw IOException("Unexpected response code (${response.code()}) from ords token refresh")
                }
            }
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
    }

    private fun buildRequest() = Request.Builder()
        .url(HttpUrl.parse(arenaOrdsUrl)!!.newBuilder().addPathSegments("arena/api/oauth/token").build())
        .header(HttpHeaders.CACHE_CONTROL, "no-cache")
        .header(HttpHeaders.AUTHORIZATION, basicAuth)
        .post(
            RequestBody.create(
                MediaType.get("application/x-www-form-urlencoded"),
                "grant_type=client_credentials"
            )
        )
        .build()


    private fun buildRequestGcp() = Request.Builder()
        .url(HttpUrl.parse(arenaOrdsUrl)!!.newBuilder().addPathSegments("arena/token").build())
        .header(HttpHeaders.CACHE_CONTROL, "no-cache")
        .header(HttpHeaders.AUTHORIZATION, "Bearer ${proxyTokenProvider()}")
        .header("Downstream-Authorization", basicAuth)
        .post(
            RequestBody.create(
                MediaType.get("application/x-www-form-urlencoded"),
                "grant_type=client_credentials"
            )
        )
        .build()

    private fun tokenIsSoonExpired(): Boolean {
        return tokenCache == null || timeToRefresh().isBefore(LocalDateTime.now())
    }

    private fun timeToRefresh(): LocalDateTime {
        return tokenCache!!.time.plus(
            (
                    tokenCache!!.ordsToken.expiresIn - MINIMUM_TIME_TO_EXPIRE_BEFORE_REFRESH).toLong(),
            ChronoUnit.SECONDS
        )
    }

    private class TokenCache(val ordsToken: OrdsToken) {
        val time: LocalDateTime = LocalDateTime.now()

    }

    data class OrdsToken(
        @JsonAlias("access_token")
        val accessToken: String,

        @JsonAlias("token_type")
        val tokenType: String,

        @JsonAlias("expires_in")
        val expiresIn: Int = 0
    )

    companion object {
        private const val MINIMUM_TIME_TO_EXPIRE_BEFORE_REFRESH = 60

        private const val ARENA_ORDS_CLIENT_ID_PROPERTY = "ARENA_ORDS_CLIENT_ID"
        private const val ARENA_ORDS_CLIENT_SECRET_PROPERTY = "ARENA_ORDS_CLIENT_SECRET"
        private val username = requireProperty(ARENA_ORDS_CLIENT_ID_PROPERTY)
        private val password = requireProperty(ARENA_ORDS_CLIENT_SECRET_PROPERTY)
        private val basicAuth = "Basic ${Base64.getEncoder().encodeToString("$username:$password".toByteArray())}"
    }
}